### PR TITLE
permission system

### PR DESCRIPTION
**Pull request by stevezhang1999**
### summary
1. uses additional module， django-guardian
  ```
pip install django-guardian
```

2. now can assign permissions **per project, per user**. currently 6 per object permissions are available.

3. how to use the new API: 

    ```python
      # Or justt do this in the admin site, see below
      assign_perm('view_project', joe, project)
      
      joe.has_perm('view_project', project)
      ```
4. now most of the comments are written in Google style.
#### per object permissions: 6 permissions
- read_project:
- edit_annotation
- edit_text
- edit_project
- remove_project
- new_project

#### default django permissions:
- add_project
- view_project

### what's in example databse:
example-project-1, owner is normal1
example-project-2, owner is normal2
normal 1 has "read_project" permission on example-project-2

### To assign permission per object:

1. click the project
     

2. click "Object Permissions" in the upper-right corner

3. Enter exact user name to find or edit users already modified before

4. Assign permissions and ignore default ones.

image
image

√: API tested
√√: API tested + tested in the system
×: not complete

### TODO:

- User without permission to add annotation could still try to add, which may lead to an error. The add button should be disabled.
- Now there is no convenient way to combine per object permissions.

### CODE REVIEW

File: simple-backend/nlpviewer_backend/handlers/project.py [hunk 1]
```diff
@@ -2,63 +2,104 @@
...
+def list_user_projects(request):
+    """
+    list all projects of the current user.
+    """
+    
+    projects_read = list(get_objects_for_user(request.user, 'nlpviewer_backend.read_project').all().values('id', 'name'))
+    projects_user = list(request.user.projects.all().values('id', 'name'))
+
+    projects_list = projects_user + projects_read
+    print(projects_list)
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/handlers/project.py:36]: Can remove these `print` statements

File: simple-backend/nlpviewer_backend/handlers/project.py [hunk 2]
```diff
@@ -2,63 +2,104 @@
...
     return JsonResponse(docJson, safe=False)
 
-
 @require_login
 def query(request, project_id):
+    """
+    query a project by id.
+    accessible for users with 'read_project' permission and the owner of the project.
+    """
+    project = fetch_project_check_perm(project_id, request.user, "nlpviewer_backend.read_project")
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/handlers/project.py:86]: What happens here when you got a `404` from the function?

**Review by hunterhector — COMMENTED** [review]
I think the overall PR should be OK. But I need to try them out before merging. 

There are a few small comments. And one more thing, could you turn your screenshots into a user guide and commit it with the PR?

File: simple-backend/nlpviewer_backend/handlers/project.py (hunk 2 above)
Comment (stevezhang1999) [reply simple-backend/nlpviewer_backend/handlers/project.py:86]: > What happens here when you got a `404` from the function?
As defined in utils.py, the function may return a 403 or 404. As a result, the current view function will not continue and its response is a 403 or 404.

File: simple-backend/nlpviewer_backend/handlers/project.py (hunk 1 above)
Comment (stevezhang1999) [reply simple-backend/nlpviewer_backend/handlers/project.py]: They are removed now

**Comment by stevezhang1999** [comment]
> I think the overall PR should be OK. But I need to try them out before merging.
> 
> There are a few small comments. And one more thing, could you turn your screenshots into a user guide and commit it with the PR?

OK, I will do it soon. Contact me if there are problems.

**Review by brocknichol — DISMISSED** [review]
The introduction of django-guardian dependency seems like a clean way to add this functionality; I think these changes look good.

File: User Guide.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
+# User Guide
+
+An example database is provided for users to get familiar with Stave. [Go to example](#example-database)
+
+## Permission System
+
+This chapter is to familiarize users with the permission system in Stave.
+
+With permission system, the user will be able to assign permission **per project**, **per project**. For example, user Joe could be allowed to edit project A but not project B.
```
[the remark this answers is not in the document]
Comment (stevezhang1999) [reply User Guide.md:9]: Thanks, I made a typo.

File: User Guide.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+This chapter is to familiarize users with the permission system in Stave.
+
+With permission system, the user will be able to assign permission **per project**, **per project**. For example, user Joe could be allowed to edit project A but not project B. 
+
+
+
+### Preparation (will be moved into README when merged)
+
+  ```
+pip install django-guardian
```
[the remark this answers is not in the document]
Comment (stevezhang1999) [reply User Guide.md:16]: Yes!  I think a requiremets.txt will be good and we may need it later.

**Review by hunterhector — APPROVED** [review]
I am re-approving this since Brock goes over the main code. New commits simply update the readme and resolve some conflicts.

**Review by brocknichol — APPROVED** [review]
Resolutions look great. Approving

**Pull request merged**